cephadm/box: Cephadm Docker in Docker dev box

pereman2 opened the pull request:
Redo of #41325. Will run teuthology.

**todo**
- [x] Remove bootstrap.sh and child scripts.

## Installation
* Install [docker-compose](https://docs.docker.com/compose/install/)
* Build frontend:
```sh
cd src/pybind/mgr/dashboard/frontend
sudo chown -R $(id -un) ./{dist,node_modules,src/environments}
npm ci
npm run build --prod
cd -
```
* Start the bootstrap script:
```sh
# at first you'll need to get the required images
cd src/cephadm/box
sudo ./box.py cluster setup
# with the required images you can now run bootstrap with: 
sudo ./box.py cluster start --osds 3 --hosts 3 # default is 1 osd 1 host
```

* Stop the cluster:
```sh
./box.py cluster down
```

# after running bootstrap.sh you can check the ips of each container with:
./box.py cluster list

# more info:
```
./box.py --help
```

Signed-off-by: Pere Diaz Bou <pdiazbou@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

sebastian-philipp commented:
no luck for me. in any case, I had this changes:

```
diff --git a/src/cephadm/box/bootstrap.sh b/src/cephadm/box/bootstrap.sh
index 41ad55429b..03892a76ef 100755
--- a/src/cephadm/box/bootstrap.sh
+++ b/src/cephadm/box/bootstrap.sh
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -x
+
 OSDS=1
 HOSTS=0
 SKIP_LOOP=0
@@ -35,6 +37,7 @@ case $1 in
         ;;
     --update-box-image)
        echo Updating box image
+       rm -f loop-images/diskloop*.img
        docker build -t cephadm-box -f Dockerfile .
         ;;
     --update-ceph-image) # TODO remove when ceph-ci updated
diff --git a/src/cephadm/box/setup_loop.sh b/src/cephadm/box/setup_loop.sh
index 661715e803..730216fe5e 100755
--- a/src/cephadm/box/setup_loop.sh
+++ b/src/cephadm/box/setup_loop.sh
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-set -e
+set -ex
 
 function create_loops() {
 
diff --git a/src/cephadm/box/setup_ssh.sh b/src/cephadm/box/setup_ssh.sh
index 5b81c2cbf6..03018673ae 100755
--- a/src/cephadm/box/setup_ssh.sh
+++ b/src/cephadm/box/setup_ssh.sh
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-set -e
+set -ex
 
 function run-sshd() {
        echo "Creating sshd server on $(hostname):$(hostname -i)"
```

alfonsomthd commented:
@pereman2 @epuertat I added in the installation description the steps to build forntend, otherwise you may not able to log in (if it's not built or it's outdated). The question is: should this script take care of building it?
AFAIK if the /dist folder does not exists dashboard module can crash...

alfonsomthd commented on src/cephadm/box/bootstrap.sh:
```diff
@@ -0,0 +1,114 @@
+	# not find lvs
+	docker-compose down
+	DCFLAGS="-f docker-compose.yml"
+        if [[ -n /sys/fs/cgroup/cgroup.controllers ]]; then
+            DCFLAGS+=" -f docker-compose.cgroup1.yml"
+        fi
```
```suggestion
    if [[ ! -f /sys/fs/cgroup/cgroup.controllers ]]; then
        DCFLAGS+=" -f docker-compose.cgroup1.yml"
    fi
```

pereman2 commented:
@sebastian-philipp There is an error when creating nfs services with cgroups v2:
```
Nov 05 16:05:06 6d995b2ae06f bash[1993]: debug 2021-11-05T16:05:06.789+0000 7f34e4efc700  0 [cephadm ERROR cephadm.serve] Failed to apply nfs.a spec NFSServiceSpec({'placement': PlacementSpec(hosts=[HostPlacementSpec(hostname='6d995b2ae>
Nov 05 16:05:06 6d995b2ae06f bash[1993]: Traceback (most recent call last):
Nov 05 16:05:06 6d995b2ae06f bash[1993]:   File "/lib64/python3.6/subprocess.py", line 425, in run
Nov 05 16:05:06 6d995b2ae06f bash[1993]:     stdout, stderr = process.communicate(input, timeout=timeout)
Nov 05 16:05:06 6d995b2ae06f bash[1993]:   File "/lib64/python3.6/subprocess.py", line 863, in communicate
Nov 05 16:05:06 6d995b2ae06f bash[1993]:     stdout, stderr = self._communicate(input, endtime, timeout)
Nov 05 16:05:06 6d995b2ae06f bash[1993]:   File "/lib64/python3.6/subprocess.py", line 1535, in _communicate
Nov 05 16:05:06 6d995b2ae06f bash[1993]:     self._check_timeout(endtime, orig_timeout)
Nov 05 16:05:06 6d995b2ae06f bash[1993]:   File "/lib64/python3.6/subprocess.py", line 891, in _check_timeout
Nov 05 16:05:06 6d995b2ae06f bash[1993]:     raise TimeoutExpired(self.args, orig_timeout)
Nov 05 16:05:06 6d995b2ae06f bash[1993]: subprocess.TimeoutExpired: Command '['ganesha-rados-grace', '--cephconf', '/tmp/mgr-grace-confb7xnmix9', '--userid', 'mgr.nfs.grace.nfs.a', '--pool', '.nfs', '--ns', 'a', 'add', 'nfs.a.0']' timed>
Nov 05 16:05:06 6d995b2ae06f bash[1993]: During handling of the above exception, another exception occurred:
Nov 05 16:05:06 6d995b2ae06f bash[1993]: Traceback (most recent call last):
Nov 05 16:05:06 6d995b2ae06f bash[1993]:   File "/lib64/python3.6/subprocess.py", line 863, in communicate
Nov 05 16:05:06 6d995b2ae06f bash[1993]:     stdout, stderr = self._communicate(input, endtime, timeout)
Nov 05 16:05:06 6d995b2ae06f bash[1993]:   File "/lib64/python3.6/subprocess.py", line 1535, in _communicate
Nov 05 16:05:06 6d995b2ae06f bash[1993]:     self._check_timeout(endtime, orig_timeout)
Nov 05 16:05:06 6d995b2ae06f bash[1993]:   File "/lib64/python3.6/subprocess.py", line 891, in _check_timeout
Nov 05 16:05:06 6d995b2ae06f bash[1993]:     raise TimeoutExpired(self.args, orig_timeout)
Nov 05 16:05:06 6d995b2ae06f bash[1993]: subprocess.TimeoutExpired: Command '['ganesha-rados-grace', '--cephconf', '/tmp/mgr-grace-confb7xnmix9', '--userid', 'mgr.nfs.grace.nfs.a', '--pool', '.nfs', '--ns', 'a', 'add', 'nfs.a.0']' timed>
```
has this error ever happened to you?

pereman2 commented on src/cephadm/cephadm:
```diff
@@ -5059,24 +5130,77 @@ def command_ceph_volume(ctx):
-        privileged=True,
-        volume_mounts=mounts,
-    )
+    # If ceph-volume creates osd data directories which won't be persisted
+    # so we use a tmp dir for that.
+    # FIXME: probably we can use /var/lib/ceph/{fsid}?
+    with tempfile.TemporaryDirectory() as tmp_osd_dir:
+        # match /var/lib/ceph/osd/ dir permissions
+        os.chown(tmp_osd_dir, uid, gid)
+        os.chmod(tmp_osd_dir, 0o755)
+        # store newly created osds here
+        mounts[tmp_osd_dir] = '/var/lib/ceph/osd/:z'
```
@sebastian-philipp @adk3798

Hey guys, I need some of your help in this problem when creating osds. I've gathered some of the things
that I've tried and listed them to discuss them:

_Different approaches to create osds with cephadm binary_

* **ceph orch daemon add osd**
  This approach doesn't seem to work neither in kcli or the box. There is 
  a problem when trying to deploy the osd since the new osd container tries to execute
  ceph-osd with a data directory, where there aren't any of the files created with
  ceph-volume because the data directory is a tmpfs mount. 
  This happens with loopback devices and with typical devices.

* **ceph orch apply** 
    This command works with filters as far as I know. To make this work it would involve
    adding logic to understand either loopback devices or lv from loopback devices
    but, I don't like this that much because loopback devices are used for the sole purpose
    of debugging.

* **cephadm ceph-volume prepare/create/...**
    First of all when calling cephadm ceph-volume to create osds, the container is missing 
    bootstrap-osd keyring so it cannot create osds. Furthermore, when creating a osd I think 
    that the block device (lv of loopback device) is formatted correctly but, the same as
    `ceph orch daemon add osd` happens where without a data directory with the expected files
    we cannot `cephadm deploy --name osd.X`.
    
    In this part of the code you'll find my proposal. If an osd was created with `cephadm ceph-volume` why
    not just map `/var/lib/ceph/osd/ceph-X`(in container) -> `/var/lib/ceph/{fsid}/osd.X`. To sum up,
    copy the osd data directory to the cephadm host and make it usable.
    
    So to deploy it would look like: 
    `cephadm ceph-volume lvm prepare --data /dev/vg1/lv0 --no-tmpfs --no-systemd`
    and then `cephadm deploy --name osd.0`.

* **making data dir manually**
   This would involve to stop messing around with cephadm binary but I haven't made it work completly.

pereman2 commented:
@adk3798 @sebastian-philipp I've updated the PR and rn I can add osds because before there was some weird behaviour with docker's storage driver. Please it would be nice to see if it works for you.

I changed the docs here https://github.com/ceph/ceph/pull/43580/files# , with crgroups 2 cephadm takes a while to deploy services and hangs when deploying osds, but if you wait and the run `box osd deploy --vg vg1` it works.

pereman2 commented:
jenkins test dashboard cephadm

pereman2 commented:
https://shaman.ceph.com/builds/ceph/wip-pereman2-testing-2021-11-01-2310/741ff7f892fac0887ff005c8f06d37dae9b3a8a8/

pereman2 commented:
https://shaman.ceph.com/builds/ceph/wip-pereman2-testing-2021-11-01-2310/2efd042fa71c01102c6fbb6b25564a10e8503cf2/ 
again :/

avanthakkar reviewed (approved):
Tested locally, LGTM!

phlogistonjohn commented on ceph.spec.in:
```diff
@@ -464,6 +464,8 @@ Summary:        Utility to bootstrap Ceph clusters
 Requires:       lvm2
 Requires:       python%{python3_pkgversion}
+Requires:       openssh-server
+Requires:       which
```
Please avoid adding a dependency on which. It's not standard or strictly needed.
https://lwn.net/Articles/874049/

epuertat replied:
@phlogistonjohn this was needed because otherwise `cephadm bootstrap` failed in the CentOS 8 container with this `which: command not found` error. I initially thought this was caused by `shutil.which` but reviewed the [source code](https://github.com/python/cpython/blob/3.6/Lib/shutil.py) and the implementation doesn't use the `which` command, so not sure who's actually calling it there.

pereman2 commented:
https://pulpito.ceph.com/pdiazbou-2022-01-03_14:43:11-rados:mgr-wip-pereman2-testing-2021-11-01-2310-distro-basic-smithi/

sebastian-philipp commented:
I just scheduled https://pulpito.ceph.com/swagner-2022-01-03_16:19:45-orch:cephadm-wip-pereman2-testing-2021-11-01-2310-distro-default-smithi/ for the orch/cephadm suite!

pereman2 commented:
@sebastian-philipp cool! Tests look great.

pereman2 commented:
I removed 2 bash files, irrelevant to the tests.

pereman2 commented:
jenkins test make check

alfonsomthd commented:
jenkins test make check

alfonsomthd commented:
jenkins test windows
